pkg/upstream/cluster/loadbalancer.go

```diff
@@ -624,3 +626,144 @@ func (lb *reqRoundRobinLoadBalancer) IsExistsHosts(metadata api.MetadataMatchCri
 func (lb *reqRoundRobinLoadBalancer) HostNum(metadata api.MetadataMatchCriteria) int {
 	return lb.hosts.Size()
 }
+
+type shortestResponseLoadBalancer struct {
+	hosts types.HostSet
+	rand  *rand.Rand
+	mutex sync.Mutex
+
+	wrrLoadBalancer types.LoadBalancer
+	fallback        bool
+
+	choice uint32
+}
+
+func newShortestResponseLoadBalancer(info types.ClusterInfo, hosts types.HostSet) types.LoadBalancer {
+	fallback := false
+	hosts.Range(func(host types.Host) bool {
+		// Check whether `UpstreamRequestDuration` is enabled,
+		// or fallback to WRR because all hosts have same duration.
+		if _, ok := host.HostStats().UpstreamRequestDuration.(gometrics.NilHistogram); ok {
+			fallback = true
+			return false
+		}
+		return true
+	})
+
+	lb := &shortestResponseLoadBalancer{
+		hosts:           hosts,
+		rand:            rand.New(rand.NewSource(time.Now().UnixNano())),
+		choice:          default_choice,
+		wrrLoadBalancer: newWRRLoadBalancer(info, hosts),
+		fallback:        fallback,
+	}
+	return lb
+}
+
+func (lb *shortestResponseLoadBalancer) ChooseHost(context types.LoadBalancerContext) types.Host {
+	if lb.fallback {
+		return lb.wrrLoadBalancer.ChooseHost(context)
+	}
+
+	hs := lb.hosts
+	total := hs.Size()
+
+	if total == 0 {
+		return nil
+	}
+	if total == 1 {
+		host := hs.Get(0)
+		if host.Health() {
+			return host
+		}
+		return nil
+	}
+
+	var candidate types.Host
+
+	// If `total` is less than or equal to `choice`, we can iterate over all elements directly.
+	if total <= int(lb.choice) {
+		log.DefaultLogger.Debugf("[lb][shortest_response] total %d <= choice %d, using iterate", total, lb.choice)
+		candidate = lb.iterateChoose()
+	} else {
+		log.DefaultLogger.Debugf("[lb][shortest_response] total %d > choice %d, using random", total, lb.choice)
+		candidate = lb.randomChoose()
+		if candidate == nil {
+			log.DefaultLogger.Warnf("[lb][shortest_response] no host chosen after %d choice, fallback to WRR", lb.choice)
+			return lb.wrrLoadBalancer.ChooseHost(context)
+		}
+	}
+
+	return candidate
+}
+
+func (lb *shortestResponseLoadBalancer) IsExistsHosts(_ api.MetadataMatchCriteria) bool {
+	return lb.hosts.Size() > 0
+}
+
+func (lb *shortestResponseLoadBalancer) HostNum(_ api.MetadataMatchCriteria) int {
+	return lb.hosts.Size()
+}
+
+func (lb *shortestResponseLoadBalancer) iterateChoose() types.Host {
+	var candidate types.Host
+	var candidateScore float64
+
+	lb.hosts.Range(func(temp types.Host) bool {
+		if !temp.Health() {
+			return false
+		}
+
+		tempScore := shortestResponseScore(temp)
+		if candidate == nil || (tempScore < candidateScore || (tempScore == candidateScore && (temp.Weight() >= candidate.Weight()))) {
```

Comment:
Maybe we can ignore the Weight and just use `tempScore < candidateScore`? Using the weight may be misleading as unlike other balancers, the weight here is just a priority for nodes that have the same score.

Reply:
When the user configures the weight, it must be intentional, and the weight should compound the distribution ratio as much as possible. Maybe I should use it as a weighting factor instead of judging when the scores are the same

Reply:
Weight is not so useful in shortestResponse or other similar response time driven balancers (like EWMA).
These balancers are considered self-balancing without giving a particular weight. For example, different sizes of nodes can have the number of requests match their sizes by making the response time equal.
Another issue is that P2C requires each host to have equal weight.

Reply:
You are right, but most implemention of Least Outstanding Request (or Least Active Request) which using P2C alse supports weighted. I don't think it's a problem, if user doesn't want to make it weighted, just keeps it to 1.

Reply:
> Weight is not so useful in shortestResponse or other similar response time driven balancers (like EWMA).

In fact, I am currently implementing the new branch (https://github.com/jizhuozhi/mosn/tree/ewma_balancer) at the fork of this branch to implement the shortest response time load balancing algorithm based on EWMA, and verify its effectiveness through repeated chaos tests.

In the latest fork I'm trying to incorporate response success rates for smarter traffic distribution.

For example, in a mixed deployment scenario, 1C, 2C, 4C, etc. exist at the same time. The slot-based concurrency control algorithm will manage the remaining available cores. When the core 1C server is overloaded and fail-fast, the load balancing based on response time Algorithms may mistakenly think this is best instead of considering servers with 2C or 4C (since they have a lot of active connections). But if we factor the response success rate into the calculation, we can avoid this from happening, because it knows that it is due to fast failure rather than it is really fast. 

Taking weight as a factor is also out of this consideration. Although we are randomly selecting hosts because of P2C, from a global perspective, the distribution ratio is in line with the weight ratio, which can also avoid this situation from happening.

Reply:
> You are right, but most implemention of Least Outstanding Request (or Least Active Request) which using P2C alse supports weighted.

Could you provide some examples? AFAIK, envoy supports both P2C and weight, but it only uses P2C for unweight situation:
https://github.com/envoyproxy/envoy/blob/db6435ac397b6bb43a212ad83bfdfd425de5ff93/source/common/upstream/load_balancer_impl.cc#L1054
And ingress-nginx supports EWMA via P2C, but it doesn't support using weight in it.

The problem about using P2C with weight is that the P2C algorithm requires each node to have the same weight. For example, considering we have 100 nodes, one of them has 100 weight, and the rest have 1 weight. By using P2C, the 100 weight node only has almost 2% to be chosen to compare. So even if it is chosen each time after comparison, only 2% of requests will be forwarded to the node, which is much less than the ratio of weight (100 / (100 + 99)).

Anyway, this is beyond the topic of this PR.

Reply:
> You are right, but most implemention of Least Outstanding Request (or Least Active Request) which using P2C alse supports weighted. I don't think it's a problem, if user doesn't want to make it weighted, just keeps it to 1.

BTW, we are discussing the weight in shortest response loadbalancer. Whether an implementation of Least Outstanding Request (or Least Active Request) uses weight or not is not so relative to the topic - as the Least Outstanding Request (or Least Active Request) is another loadbalancer.

Reply:
Hello, @muyuan0 . There is an example that NGINX supports weighted connection when using P2C (named Random2 in NGINX). 
See: https://github.com/nginx/nginx/blob/master/src/http/modules/ngx_http_upstream_random_module.c#L365-L373

Reply:
If you're interested, we can discuss it in a new PR (#2274 ) where more optimizations are done over the current implementation. And I am doing chaos testing in various scenarios, and I will attach the test report to the PR later, welcome to discuss and make suggestions :)

Reply:
> Hello, @muyuan0 . There is an example that NGINX supports weighted connection when using P2C (named Random2 in NGINX). See: [nginx/nginx@`master`/src/http/modules/ngx_http_upstream_random_module.c#L365-L373](https://github.com/nginx/nginx/blob/master/src/http/modules/ngx_http_upstream_random_module.c?rgh-link-date=2023-03-30T14%3A43%3A00Z#L365-L373)

I am afraid you make a mistake here:
Nginx doesn't implement the standard P2C algorithm, at least it doesn't implement the P2C as this PR does.
Here it uses O(logn) to find the random node:
https://github.com/nginx/nginx/blob/dfe70f74a3558f05142fb552cea239add123d414/src/http/modules/ngx_http_upstream_random_module.c?rgh-link-date=2023-03-30T14%3A43%3A00Z#L407
It can't use O(1) to find the random node, otherwise, it will require O(n*weight) memory footprint.
However, this PR uses O(1) algorithm to select the node.

Reply:
Another issue, as I pointed out in https://github.com/mosn/mosn/pull/2253#discussion_r1153315997, 
> BTW, we are discussing the weight in shortest response loadbalancer. Whether an implementation of Least Outstanding Request (or Least Active Request) uses weight or not is not so relative to the topic - as the Least Outstanding Request (or Least Active Request) is another loadbalancer.

you are talking about the Least Connections balancer. The link to the Nginx code is nothing about the shortest response balancer.

Reply:
What I want to explain is that both Least Request and Shortest Response measure the processing capacity of the server to select the server with the fastest processing request, so their optimization ideas are interoperable.

The Least Request is based on the fact that water is injected into multiple leaking buckets at the same rate, and they are observed at the same time at a certain moment. The probability of the remaining water in the bucket is small is that the flow rate is fast. But when the outflow rate is faster than the injection rate, so that every observation is almost empty, then the number of connections at this time will not be able to support the decision. Although it is observed that the bucket is almost empty, the difference in the outflow rate still exists. 

On this basis, Shortest Response tries to collect samples at the end of the request and form a characteristic value accumulated by all historical requests. Although the characteristic value can fully reflect the server's processing capacity, we still need to avoid the last straw that crushes the camel's camel. This is the same as many EWMA algorithms use the number of connections for weighting (including linkerd and this implementation), or the response time is the thing called the weight of the number of connections.

As for nginx, yes, nginx does not implement standard P2C, it uses a weighted random algorithm. In fact, the concept of weight is always emphasized in the implementation of NGINX (just like RR is actually SWRR), it always hopes Give users a chance to control traffic, as I think so too. This is very common in actual deployment scenarios. Servers of different specifications are mixed in the same cluster and provide external services at the same time. There are many reasons for this. What I have experienced is that cloud service providers do not have enough resources in a certain availability zone Instances of the same specifications, then we can only purchase instances of different specifications and adjust the distribution ratio of traffic through weights

Reply:
As for those users who do not want to be affected by the weight, it is very simple, just configure all nodes to have the same weight or not (the default is 1). 

Since the user configures the weight, there must be an intention to distribute it according to the weight. As a load balancing algorithm, it should satisfy this distribution method as much as possible. This is my idea.

Reply:
> As for those users who do not want to be affected by the weight, it is very simple, just configure all nodes to have the same weight or not (the default is 1).

It is not as simple as it looks. We need to change the code according to the support of weight.
As I mentioned in https://github.com/mosn/mosn/pull/2253#discussion_r1153454010, if weight is needed to support, we can't have O(1) random choosing. I would recommend implementing P2C only for unweight branch, so in most case (the same weight is used) there is no need to use a complex O(logn) lookup.

```
	// If `total` is less than or equal to `choice`, we can iterate over all elements directly.
	if total <= int(lb.choice) {
		candidate = lb.iterateChoose()
	} else { // should also consider the weight here
		candidate = lb.randomChoose()
```

Reply:
The weight factor is finally retained, but it is not used to distribute traffic according to the weight, but as a trade-off of processing capacity, just like the weighted least active request, it is an estimate of the processing time of active requests. In other words, called the number of active requests per unit resource